source/common/network/listener_filter_buffer_impl.cc

```diff
@@ -0,0 +1,70 @@
+#include "source/common/network/listener_filter_buffer_impl.h"
+
+#include <string.h>
+
+namespace Envoy {
+namespace Network {
+
+uint64_t ListenerFilterBufferImpl::copyOut(Buffer::Instance& buffer, uint64_t length) {
+  auto size = std::min(length, data_size_);
+  if (size == 0) {
+    return 0;
+  }
+  buffer.drain(buffer.length());
+  uint8_t* raw_mem = static_cast<uint8_t*>(buffer_->frontSlice().mem_);
+  buffer.add(raw_mem + drained_size_, size);
+  return size;
+}
+
+uint64_t ListenerFilterBufferImpl::drain(uint64_t length) {
+  auto size_to_drain = std::min(length, data_size_);
+  // It doesn't drain the data from buffer directly until drain the data
+  // from actual socket.
+  drained_size_ += size_to_drain;
+  data_size_ -= size_to_drain;
+  return size_to_drain;
+}
+
+bool ListenerFilterBufferImpl::drainFromSocket() {
+  // Since we want to drain the data from the socket, so a
+  // temporary buffer need here.
+  std::unique_ptr<uint8_t[]> buf(new uint8_t[drained_size_]);
```

Comment:
I would use https://github.com/abseil/abseil-cpp/blob/master/absl/container/fixed_array.h here. I'm guessing in some cases this might get stack allocated. Also generally safer.

Reply:
got it

Reply:
Sorry, just found I can't use the fixed array, since it need a template parameter for the size of array. But the size is decided in the runtime. (how much data drained by filter).